### PR TITLE
Solution for task # 3

### PR DESCRIPTION
**Pull request by MorozVladislav**
I'm an artist. I see so)

### CODE REVIEW

File: solutions/task_3.py [hunk 1]
```diff
@@ -0,0 +1,105 @@
+#!/usr/bin/env python3
+"""
+TASK: Implement decorators 'add_2', 'add_n', 'cached' and class 'Contact' to perform code below without errors.
+"""
+import time
+from datetime import datetime, timedelta
+
+
+def add_2(func):
+    def wrapper(a, b):
```
Comment (dmitry-moroz) [inline solutions/task_3.py]: 1) Why are you passing to ```wrapper``` the same args ```(a, b)``` as in function ```multiply_1```? What if arguments for function have been changed? You have to change it in two places. Do you think that it's ok?
2) I suggest to name wrapped function "wrappe**d**".

File: solutions/task_3.py [hunk 2]
```diff
@@ -0,0 +1,105 @@
...
+        return func(a, b) + 2
+    return wrapper
+
+
+def add_n(n):
+    def wrapper_1(func):
+        def wrapper_2(a, b):
```
Comment (dmitry-moroz) [inline solutions/task_3.py]: 1) The same as above about arguments.
2) I suggest to rename:
```
wrapper_1 -> wrapper
wrapper_2 -> wrapped
```

File: solutions/task_3.py [hunk 3]
```diff
@@ -0,0 +1,105 @@
...
+            return func(a, b) + n
+        return wrapper_2
+    return wrapper_1
+
+
+CACHE = {}
```
Comment (dmitry-moroz) [inline solutions/task_3.py]: Define global variables at the top of file.

File: solutions/task_3.py [hunk 4]
```diff
@@ -0,0 +1,105 @@
...
+            CACHE[func][str(args)+str(kwargs)][0] = func(*args, **kwargs)
+            CACHE[func][str(args)+str(kwargs)][1] = datetime.now()
+            return CACHE[func][str(args)+str(kwargs)][0]
+        return wrapper_2
+    return wrapper_1
+
+
+class Contact(object):
+    def __init__(self, phone, operator):
+        self._phone_ = phone
```
Comment (dmitry-moroz) [inline solutions/task_3.py]: ```self._phone```
Why do you need second underscore? One enough.

File: solutions/task_3.py [hunk 5]
```diff
@@ -0,0 +1,105 @@
...
+        self.operator = operator
```
Comment (dmitry-moroz) [inline solutions/task_3.py]: Please implement ```operator``` like a ```@property``` too.

File: solutions/task_3.py [hunk 6]
```diff
@@ -0,0 +1,105 @@
...
+
+    @property
+    def phone(self):
+        return "{} {} {}".format("".join(self._phone_[:4]), "".join(self._phone_[4:]), self.operator)
```
Comment (dmitry-moroz) [inline solutions/task_3.py]: 1) Get rid of this ```"{} {} {}"```. Use ```join```.
2) What if code is longer than 4 digits? Phone number length is always = 7.

File: solutions/task_3.py [hunk 7]
```diff
@@ -0,0 +1,105 @@
...
+
+CACHE = {}
+
+
+def cached(cache_life_time):
+    def wrapper_1(func):
+        CACHE[func] = {}
+
+        def wrapper_2(*args, **kwargs):
+            if str(args) + str(kwargs) not in CACHE[func].keys():
```
Comment (dmitry-moroz) [inline solutions/task_3.py]: 1) Why are you calculation this key ```str(args) + str(kwargs)``` all the time?! Why you can't do it once?
2) ```not in CACHE[func].keys()``` you don't need ```keys()``` it's enough: ```not in CACHE[func]```

File: solutions/task_3.py [hunk 8]
```diff
@@ -0,0 +1,105 @@
...
+                CACHE[func][str(args) + str(kwargs)] = [None, None]
+                CACHE[func][str(args) + str(kwargs)][0] = func(*args, **kwargs)
+                CACHE[func][str(args) + str(kwargs)][1] = datetime.now()
```
Comment (dmitry-moroz) [inline solutions/task_3.py]: **Wtf is going here?!?!** (lines 32-35)
```
result = func(*args, **kwargs)
CACHE[func][key] = (result, datetime.now())
return result
```

File: solutions/task_3.py [hunk 9]
```diff
@@ -0,0 +1,105 @@
...
+                return CACHE[func][str(args) + str(kwargs)][0]
+            if (datetime.now() - CACHE[func][str(args)+str(kwargs)][1]).total_seconds() < cache_life_time:
```
Comment (dmitry-moroz) [inline solutions/task_3.py]: To not use ```.total_seconds()``` use ```time.time()``` to get Unix time, datetime lib is for more complex things.

**Review by dmitry-moroz — CHANGES_REQUESTED** [review]

File: solutions/task_3.py [hunk 10]
```diff
@@ -0,0 +1,105 @@
...
+                return CACHE[func][str(args)+str(kwargs)][0]
+            CACHE[func][str(args)+str(kwargs)][0] = func(*args, **kwargs)
```
Comment (dmitry-moroz) [inline solutions/task_3.py]: Do you know construction?
```
if ... :
    ...
else:
    ...
```

File: solutions/task_3.py (hunk 1 above)
Comment (MorozVladislav) [reply solutions/task_3.py]: Agree. Fixed.

File: solutions/task_3.py (hunk 2 above)
Comment (MorozVladislav) [reply solutions/task_3.py]: Agree. Fixed.

File: solutions/task_3.py (hunk 3 above)
Comment (MorozVladislav) [reply solutions/task_3.py]: Fixed.

File: solutions/task_3.py (hunk 7 above)
Comment (MorozVladislav) [reply solutions/task_3.py]: Agree. Fixed.

File: solutions/task_3.py (hunk 8 above)
Comment (MorozVladislav) [reply solutions/task_3.py]: Sorry. I was young and drunk.

File: solutions/task_3.py (hunk 4 above)
Comment (MorozVladislav) [reply solutions/task_3.py]: Fixed.

File: solutions/task_3.py (hunk 5 above)
Comment (MorozVladislav) [reply solutions/task_3.py]: Done.

File: solutions/task_3.py (hunk 6 above)
Comment (MorozVladislav) [reply solutions/task_3.py]: Fixed.

File: solutions/task_3.py (hunk 10 above)
Comment (MorozVladislav) [reply solutions/task_3.py]: Fixed.

File: solutions/task_3.py [hunk 11]
```diff
@@ -6,57 +6,63 @@
...
     def __init__(self, phone, operator):
-        self._phone_ = phone
-        self.operator = operator
+        self._phone = phone
+        self._operator = operator
 
     @property
     def phone(self):
-        return "{} {} {}".format("".join(self._phone_[:4]), "".join(self._phone_[4:]), self.operator)
+        return "".join("".join(self._phone[:-7]) + " " + "".join(self._phone[-7:]) + " " + self.operator)
```
Comment (dmitry-moroz) [inline solutions/task_3.py]: Omg...
Please. Tell. Me. What. This. Code. Is. Doing? What. Is. Sense. Do. You. See. Here?
How are you able to write such code?
```
Hmmm... I have a string '1234567'
How can I convert it to the same string? Idea!
I can split it into chars -> ['1', '2', '3', '4', '5', '6', '7']
And join all of them with empty string!! "".join(['1', '2', '3', '4', '5', '6', '7'])
Great! I have got the same string '1234567'!
```

File: solutions/task_3.py [hunk 12]
```diff
@@ -6,57 +6,63 @@
...
+            key = str(args) + str(kwargs)
+            if key not in CACHE[func]:
+                CACHE[func][key] = [func(*args, **kwargs), time.time()]
+                return CACHE[func][key][0]
+            if (time.time() - CACHE[func][key][1]) < cache_life_time:
+                return CACHE[func][key][0]
+            else:
+                CACHE[func][key][0] = func(*args, **kwargs)
+                CACHE[func][key][1] = time.time()
+                return CACHE[func][key][0]
```
Comment (dmitry-moroz) [inline solutions/task_3.py]: If you will improve this code (lines 39-41) you will get the same code as above:
```
CACHE[func][key] = [func(*args, **kwargs), time.time()]
return CACHE[func][key][0]
```
Would you like to improve this moment? (I suppose yes)

File: solutions/task_3.py [hunk 13]
```diff
@@ -6,57 +6,63 @@
...
-                CACHE[func][str(args) + str(kwargs)] = [None, None]
-                CACHE[func][str(args) + str(kwargs)][0] = func(*args, **kwargs)
-                CACHE[func][str(args) + str(kwargs)][1] = datetime.now()
-                return CACHE[func][str(args) + str(kwargs)][0]
-            if (datetime.now() - CACHE[func][str(args)+str(kwargs)][1]).total_seconds() < cache_life_time:
-                return CACHE[func][str(args)+str(kwargs)][0]
-            CACHE[func][str(args)+str(kwargs)][0] = func(*args, **kwargs)
-            CACHE[func][str(args)+str(kwargs)][1] = datetime.now()
-            return CACHE[func][str(args)+str(kwargs)][0]
+            key = str(args) + str(kwargs)
```
Comment (dmitry-moroz) [inline solutions/task_3.py]: Why you don't rename your ```wrapper_1``` and ```wrapper_2``` as above?
Why I have to write the same comments twice? You have to apply all the comments from PR to all the code, isn't it?

**Review by dmitry-moroz — CHANGES_REQUESTED** [review]

**Pull request merged**